File: php/day14_part1_2019.php
<?php

class Chemical {
    public $name;
    public $amount;

    public function __construct($name, $amount) {
        $this->name = $name;
        $this->amount = $amount;
    }
}

function parseChemical($s) {
    $parts = explode(" ", $s);
    return new Chemical($parts[1], intval($parts[0]));
}

function calculateOre($chem, $amount, $reactions, $ingredients, &$surplus) {
    if ($chem == "ORE") {
        return $amount;
    }

    if (isset($surplus[$chem]) && $surplus[$chem] >= $amount) {
        $surplus[$chem] -= $amount;
        return 0;
    }

    $amount -= isset($surplus[$chem]) ? $surplus[$chem] : 0;
    $surplus[$chem] = 0;
    $reaction = $reactions[$chem];
    $times = (int)(($amount + $reaction->amount - 1) / $reaction->amount);
    $ore = 0;

    foreach ($ingredients[$chem] as $ingredient) {
        $ore += calculateOre($ingredient->name, $ingredient->amount * $times, $reactions, $ingredients, $surplus);
    }

    $surplus[$chem] = $times * $reaction->amount - $amount;
    return $ore;
}

$file = fopen("input.txt", "r");
$reactions = [];
$ingredients = [];

while (($line = fgets($file)) !== false) {
    $parts = explode(" => ", trim($line));
    $output = parseChemical($parts[1]);
    $inputs = [];
    foreach (explode(", ", $parts[0]) as $input) {
        $inputs[] = parseChemical($input);
    }
    $reactions[$output->name] = $output;
    $ingredients[$output->name] = $inputs;
}

fclose($file);

$surplus = [];
echo calculateOre("FUEL", 1, $reactions, $ingredients, $surplus);

File: lua/day14_part1_2019.lua
function parseChemical(s)
    local parts = {}
    for part in s:gmatch("%S+") do
        table.insert(parts, part)
    end
    local amount = tonumber(parts[1])
    local name = parts[2]
    return {name = name, amount = amount}
end

function calculateOre(chem, amount, reactions, ingredients, surplus)
    if chem == "ORE" then
        return amount
    end

    if not surplus[chem] then
        surplus[chem] = 0
    end

    if surplus[chem] >= amount then
        surplus[chem] = surplus[chem] - amount
        return 0
    end

    amount = amount - surplus[chem]
    surplus[chem] = 0
    local reaction = reactions[chem]
    local times = math.ceil(amount / reaction.amount)
    local ore = 0

    for _, ingredient in ipairs(ingredients[chem]) do
        ore = ore + calculateOre(ingredient.name, ingredient.amount * times, reactions, ingredients, surplus)
    end

    surplus[chem] = surplus[chem] + times * reaction.amount - amount
    return ore
end

local reactions = {}
local ingredients = {}
local surplus = {}

local file = io.open("input.txt", "r")
if not file then
    print("Error opening file")
    return
end

for line in file:lines() do
    local parts = {}
    for part in line:gmatch("[^=>]+") do
        table.insert(parts, part:match("^%s*(.-)%s*$"))
    end
    local output = parseChemical(parts[2])
    local inputs = {}
    for inChem in parts[1]:gmatch("[^,]+") do
        table.insert(inputs, parseChemical(inChem))
    end
    reactions[output.name] = output
    ingredients[output.name] = inputs
end

file:close()

-- Initialize surplus for each chemical
for chem in pairs(reactions) do
    surplus[chem] = 0
end

print(calculateOre("FUEL", 1, reactions, ingredients, surplus))